includes/class-amp-theme-support.php

```diff
@@ -791,36 +791,54 @@ protected static function ensure_required_markup( DOMDocument $dom ) {
 
 	/**
 	 * Start output buffering.
+	 *
+	 * @since 0.7
+	 * @see AMP_Theme_Support::finish_output_buffering()
 	 */
 	public static function start_output_buffering() {
-		ob_start( array( __CLASS__, 'finish_output_buffering' ) );
+		ob_start();
+
+		// Note that the following must be at 0 because wp_ob_end_flush_all() runs at shutdown:1.
+		add_action( 'shutdown', array( __CLASS__, 'finish_output_buffering' ), 0 );
 	}
 
 	/**
 	 * Finish output buffering.
 	 *
-	 * @todo Do this in shutdown instead of output buffering callback? This will make it easier to debug things since printing can be done in shutdown function but cannot in output buffer callback.
+	 * @since 0.7
+	 * @see AMP_Theme_Support::start_output_buffering()
+	 */
+	public static function finish_output_buffering() {
+		$output = ob_get_clean();
```

Comment:
I would advise to avoid storing the output in a variable unless necessary, especially in the context of the output buffer. For instance the memory goes from 3018376 to 3006088 on my local install by moving `ob_get_clean()` to `self::prepare_response( ob_get_clean() );`.